pkg/step/codegen.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2020 The SQLFlow Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package step
+
+import (
+	"fmt"
+
+	pb "sqlflow.org/sqlflow/pkg/proto"
+	"sqlflow.org/sqlflow/pkg/sql/codegen/couler"
+	"sqlflow.org/sqlflow/pkg/sql/ir"
+)
+
+// Codegen represents all kinds of Code Generators,
+// such as Tensorflow, XGBoost and Couler e.g.
+type Codegen interface {
+	SQLProgram(sp ir.SQLProgram, sess *pb.Session) (string, error)
+	Train(ts *ir.TrainStmt, sess *pb.Session) (string, error)
+	Explain(es *ir.ExplainStmt, sess *pb.Session) (string, error)
+	Predict(ps *ir.PredictStmt, sess *pb.Session) (string, error)
```

Comment:
All thee methods `Train`, `Explain`, and `Predict` are expecting the same `pb.Session`. It would be a better idea to embed the session as a data member inside the implementation of interface Codegen.  So, users can write

```go
ir, e := parser.ParseUnitProgam(a_sql_stmt) // <- Let's rename ParseSingleStatement to ParseUnitProgram
if e != nil {
    return e
}

cg := NewCodegen(os.Getenv("SQLFLOW_CODEGEN"), session) // <- pass session to New
pg, e := cg.Train(ir) // <- instead of Train
if e != nil {
    return e
}

rn := NewRunner(os.Getenv("SQLFLOW_CODEGEN"), session)
rn.Run(pg)
```

The function `step.Runner.Run` could be

```go
func (rn *Runner) Run(program string) error {
    if rn.codegen != "couler" {
        cmd := exec.Cmd("python")
        cmd.Stdin = strings.NewReader(pg)
        return cmd.Run()
    } 
    return couler.Run(couler.Compile(program))
}
```